File: codechef/SEPT16/Q5.py
def getIndex(var,bb):
    if var==1 or bb==-1:return -1
    for z in range(bb,length):  # and z*z<var
        if primes[z] > var: break 
        if var%primes[z]==0:
            return z
    return -1
def update(l,r):
    for i in range(l,r+1):
        if a[i]==1 or b[i]==-1 or c[i]==1:continue
        a[i]//=c[i]
        b[i]=getIndex(a[i],b[i])
        if b[i]==-1:c[i]=1
        else:c[i]=primes[b[i]]
limit=(10**3+3)
a=[1]*limit
a[0]=a[1]=0
primes=[]
for i in range(2,limit):
    if a[i]==1:
        for j in range(i*i, limit, i):
            a[j]=0
        primes.append(i)
length=len(primes)
//print("Primes",primes,len(primes))
t=int(input())          
for e in range(t):
    n,m=map(int,input().split())
    a,b,c=[],[-1]*(n+1),[1]*(n+1)
    a=list(map(int,input().split()))
    a.insert(0,0)
    for i in range(1,n+1):
        b[i]=getIndex(a[i],0)
        if b[i]!=-1:c[i]=primes[b[i]]
        else:c[i]=1
    for i in range(m):
        tp,l,r=map(int,input().split())
        if tp==0:
            update(l,r)
        else:
            s=set(c[l:(r+1)])
            print(max(s),end=" ")
            #print("\n",a,c,b)
    print()

File: codechef/SEPT16/Q5.cpp
#include <iostream>
#include <cstdio>
#include <algorithm>
#include <vector>
#include <set>
using namespace std;
#define len v.size()

int a[100002],b[100002],pLen,n;       // Q 5
vector<int> primes;
vector<int> v;
int BSearch(int);

int get(int l,int r){
    if(v.size()==0)return 1;
    int result=1,i,x;
    i=BSearch(l);
    while(i<v.size()){
        x=v[i];
        if(x>r )break;
        if(b[x]<pLen)result=max(result,primes[b[x]]);
        else result=max(result,a[x]);
        i++;
    }
    return result;
}

void update(int l,int r){
    if(v.size()==0)return;
    int result=1,i,x;
    i=BSearch(l);
    while(i<v.size()){
        x=v[i];
        if(x>r)break;
        if(b[x]>=pLen)a[x]=1;
        else a[x]=(a[x]/primes[b[x]] );
        if(a[x]==1){
            v.erase(v.begin()+i);
        }else{
            i++;
            while((a[x]%primes[b[x]])!=0){
                b[x]++;
                if(b[x]>=pLen)break;
            }
        }
    }
//printf("v => ");for(int i=0;i<len;i++)printf("%d ",v[i]);cout<<endl;
//printf("a => ");for(int i=1;i<=n;i++)printf("%d ",a[i]);cout<<endl;
//printf("b => ");for(int i=1;i<=n;i++)printf("%d  ",b[i]);cout<<endl;
}

int BSearch(int key){
    int m,l=0,r=len-1;
    while(l<=r){
        m=(l+r)/2;
        if(key==v[m] )return m;
        else if(v[m]<key)l=m+1;
        else r=m-1;

        if(key==v[l] )return l;
        if(key==v[r] )return r;
    }
    return l;
}
void generatePrimes(){
    int i,j,k,z,x=998;
    for (i=2;i<x;i++)primes.push_back(i);
    for (i=0;i<primes.size();i++) {
        j=i+1;z=primes[i];
        k=z*z;
        while(j<primes.size()){
            if(primes[j]==k){
                primes.erase(primes.begin()+j);
                k+=z;
            }else if(k<primes[j]){
                k+=z;
            }
            else j++;
        }
    }
}

int main() {
    int t;
    scanf("%d",&t);
    generatePrimes();
    pLen=primes.size();
    while(t--){
        int m,l,r,c,ans=1;
        v.clear();
        scanf("%d%d",&n,&m);
        for(int i=1;i<=n;i++){
            scanf("%d",&a[i]);
            if(a[i]==1)continue;
            v.push_back(i);
            b[i]=0;
            while( (a[i]%primes[b[i]])!=0){
                b[i]++;
                if(b[i]>=pLen){
                    break;
                }
            }
        }
//printf("a => ");for(int i=1;i<=n;i++)printf("%d ",a[i]);cout<<endl;
//printf("b => ");for(int i=1;i<=n;i++)printf("%d  ",b[i]);cout<<endl;
        for(int i=1;i<=m;i++){
           scanf("%d %d %d",&c,&l,&r);
           if(c==0){
               update(l,r);
           }
           else {
               ans=get(l,r);
               printf("%d ",ans);
               ///printf("ans=%d \n",ans);
            }
        }printf("\n");
    }
	return 0;
}
